peps/pep-0788.rst

```diff
@@ -14,85 +14,77 @@ Post-History: `10-Mar-2025 <https://discuss.python.org/t/83959>`__,
+references that prevent an interpreter from finalizing (or more technically,
+entering a stage in which attachment of a thread state hangs).
+This allows for more structure and reliability when it comes to thread state
+management, because it forces a layer of synchronization between the
+interpreter and the caller.
+
+With this new system, there are a lot of changes needed in CPython and
+third-party libraries to adopt it. For example, in APIs that don't require
+the caller to hold an attached thread state, a strong interpreter reference
+should be passed to ensure that it targets the correct interpreter, and that
+the interpreter doesn't concurrently deallocate itself. The best example of
+this in CPython is :c:func:`PyGILState_Ensure`. As part of this proposal,
+:c:func:`PyThreadState_Ensure` is provided as a modern replacement that
+takes a strong interpreter reference.
+
+Terminology
+===========
 
-This is achieved by introducing two concepts into the C API:
+Interpreters
+------------
 
--  "Daemon" and "non-daemon" threads, similar to how it works in the
-   :mod:`threading` module.
--  Interpreter reference counts which prevent an interpreter from finalizing.
+In this proposal, "interpreter" refers to a singular, isolated interpreter
+(see :pep:`684`), with its own :c:type:`PyInterpreterState` pointer (referred
+to as an "interpreter-state"). "Interpreter" *does not* refer to the entirety
+of a Python process.
 
-In :c:func:`PyThreadState_Ensure`, both of these ideas are applied. The
-calling thread is to store a reference to an interpreter via
-:c:func:`PyInterpreterState_Hold`. :c:func:`PyInterpreterState_Hold`
-increases the reference count of an interpreter, requiring the thread
-to finish (by eventually calling :c:func:`PyThreadState_Release`) before
-beginning finalization.
+The "current interpreter" refers to the interpreter-state
+pointer on an :term:`attached thread state`, as returned by
+:c:func:`PyThreadState_GetInterpreter`.
 
-For example, creating a native thread with this API would look something
-like this:
+Native and Python Threads
+-------------------------
 
-.. code-block:: c
+This PEP refers to a thread created using the C API as a "native thread",
+also sometimes referred to as a "non-Python created thread", where a "Python
+created" is a thread created by the :mod:`threading` module.
 
-    static PyObject *
-    my_method(PyObject *self, PyObject *unused)
-    {
-        PyThread_handle_t handle;
-        PyThead_indent_t indent;
-
-        PyInterpreterState *interp = PyInterpreterState_Hold();
-        if (PyThread_start_joinable_thread(thread_func, interp, &ident, &handle) < 0) {
-            PyInterpreterState_Release(interp);
-            return NULL;
-        }
-        /* The thread will always attach and finish, because we increased
-           the reference count of the interpreter. */
-        Py_RETURN_NONE;
-    }
+Native threads are typically created by :c:func:`PyGILState_Ensure`, but more
```

Comment:
PyGILState_Ensure() doesn't create a thread. It only attachs a Python thread state to an existing thread. The sentence is misleading.